examples/hubert/dataset/hubert_dataset.py

```diff
@@ -34,65 +33,78 @@ class BucketizeSampler(BatchSampler):
 
     def __init__(
         self,
-        data_source: Dataset,
+        lengths: List[int],
         num_buckets: int,
+        min_len: int = 0,
+        max_len: Optional[int] = None,
         max_token_count: Optional[int] = None,
         batch_size: Optional[int] = None,
+        shuffle: bool = True,
     ) -> None:
+        if max_len is None:
+            max_len = max(lengths)
+
+        if not (0 <= min_len <= max_len):
+            raise AssertionError("``min_len`` should be non-negative and smaller than ``max_len``")
         if max_token_count is not None and batch_size is not None:
             raise AssertionError("The ``max_token_count`` and ``batch_size`` can't be both set.")
-        self.data_source = data_source
+        # Filter out examples which are outside the bounds of [min_len, max_len]
+        # sort to minimize gap when bucketizing.
+        self.lengths = list(sorted(length for length in lengths if min_len <= length <= max_len))
```

Comment:
The indices needs to be updated since `lengths` order is changed.
```python
length_id = [(length, i) for i, length in enumerate(lengths) if min_len <= length <= max_len]
sorted_length_id = list(sorted(length_id, key = lambda x : x[0]))
self.lengths = ...
self.indices = ...
```

Reply:
I see. So the `for i, _ in enumerate(len_list)` from earlier code should just be `for i in self.indices`?

Reply:
Maybe to make it equivalent it would be `for i, _ in zip(self.indices, self.lengths)`, but the zip and ignore is basically a no-op.

Reply:
Exactly. ``self.indices`` makes more sense. We just need to make sure the index matches the one from the original dataset.

Reply:
I have something like this:

```patch
+        filtered_lengths = torch.tensor([length for length in lengths if min_len <= length <= max_len])
+        if len(filtered_lengths) == 0:
             raise AssertionError("``lengths`` cannot be empty after filtering.")
+        self.length_idx = torch.argsort(filtered_lengths)
+        self.lengths = filtered_lengths[length_idx]
         self.max_token_count = max_token_count
         self.batch_size = batch_size
-        self.buckets = self._get_buckets(self.lengths, num_buckets, min_len, max_len)
+        self.buckets = self._get_buckets(self.reordered_lengths, self.length_idx, num_buckets, min_len, max_len)
```

where `_get_buckets`s only code change is that the enumeration is replaced by iteration over `self.indices`. Basically it's keeping track of the original place of the example in the dataset. This seems pretty close to my implementation -- sorry it's taking a bit of time for this to click. 😅

Reply:
Thanks. When we filter out the lengths, the indices already changed. So we can sort the lengths first to get the indices, then filter the lengths and indices, what do you think?

Reply:
Yup, realized that right after posting this diff. OK, I'll make these changes quickly and push it. Looks like CI failures are unrelated, too...